p2p/connectionpool/connectionpool.go

```diff
@@ -0,0 +1,190 @@
+	cp.connMutex.Lock()
+	cp.net.GetLogger().Info("new connection %v -> %v. id=%s", cp.localPub, rPub, conn.ID())
+	// check if there isn't already same connection (possible if the second connection is a Remote connection)
+	cur, ok := cp.connections[rPub.String()]
+	if ok {
+		if cur.Source() == net.Remote {
+			cp.net.GetLogger().Info("connection created by remote node while connection already exists between peers, closing new connection. remote=%s", rPub)
+		} else {
+			cp.net.GetLogger().Warning("connection created by local node while connection already exists between peers, closing new connection. remote=%s", rPub)
+		}
+		conn.Close()
+		cp.connMutex.Unlock()
+		return
+	}
+	cp.connections[rPub.String()] = conn
+	cp.connMutex.Unlock()
+
+	// update all registered channels
+	cp.pendMutex.Lock()
+	res := DialResult{conn, nil}
+	for _, p := range cp.pending[rPub.String()] {
+		p <- res
+	}
+	delete(cp.pending, rPub.String())
+	cp.pendMutex.Unlock()
+}
+
+func (cp *ConnectionPool) handleClosedConnection(conn net.Connectioner) {
+	cp.net.GetLogger().Info("connection %s was closed", conn.String())
+	cp.connMutex.Lock()
+	rPub := conn.RemotePublicKey().String()
+	cur, ok := cp.connections[rPub]
+	// only delete if the closed connection is the same as the cached one (it is possible that the closed connection is a duplication and therefore was closed)
+	if ok && cur.ID() == conn.ID() {
+		delete(cp.connections, rPub)
+	}
+	cp.connMutex.Unlock()
+}
+
+// fetch or create connection to the address which is associated with the remote public key
+func (cp *ConnectionPool) GetConnection(address string, remotePub crypto.PublicKey) (net.Connectioner, error) {
+	cp.connMutex.RLock()
+	if cp.shutdown {
+		cp.connMutex.RUnlock()
+		return nil, errors.New("ConnectionPool was shut down")
+	}
+	// look for the connection in the pool
+	conn, found := cp.connections[remotePub.String()]
+	if found {
+		cp.connMutex.RUnlock()
+		return conn, nil
+	}
+	// register for signal when connection is established - must be called under the connMutex otherwise there is a race
+	// where it is possible that the connection will be established and all registered channels will be notified before
+	// the current registration
+	cp.pendMutex.Lock()
+	_, found = cp.pending[remotePub.String()]
+	if !found {
+		// No one is waiting for a connection with the remote peer, need to call Dial
+		go func() {
+			cp.dialWait.Add(1)
+			conn, err := cp.net.Dial(address, remotePub, cp.net.GetNetworkId())
+			if err != nil {
+				cp.handleDialError(remotePub, err)
+			} else {
+				cp.handleNewConnection(remotePub, conn)
+			}
+			cp.dialWait.Done()
+		}()
+	}
+	pendChan := make(chan DialResult)
+	cp.pending[remotePub.String()] = append(cp.pending[remotePub.String()], pendChan)
+	cp.pendMutex.Unlock()
+	cp.connMutex.RUnlock()
+	// wait for the connection to be established, if the channel is closed (in case of dialing error) will return nil
+	res := <-pendChan
```

Comment:
We talked about - I think that both methods make sense. The current method has two advantages though- I passes back the error (in case of failure) and it is already implemented :)